Update vault docs for client scripts, multiple keys and misc

orthanc opened the pull request:
##### SUMMARY
Update the ansiblle-vault docs to explain the special handling of -client scripts which are passed the vault id.

Also update the ansible-playbook docs of vault to prefer --vault-id and indicate that multiple keys are suported.

The switch to vault-id is based on the main ansibe-vault docs indcating that's the preferred option. I'd particularly like confirmation that this is preferred as personally I'd argue in favour of preferring ask-vault-pass/vault-password-file over vault-id when only a single key is in use since the vault-id syntax is confusing when the vault is not named.

Fixes #43991

##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
ansible-vault

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (update_vault_docs_for_client_scripts 690f1d52ae) last updated 2018/08/11 17:32:08 (GMT +1300)
  config file = None
  configured module search path = [u'USER_HOME/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = ANSIBLE_HOME/lib/ansible
  executable location = ANSIBLE_HOME/bin/ansible
  python version = 2.7.15rc1 (default, Apr 15 2018, 21:51:34) [GCC 7.3.0]
```

orthanc committed:
Fix confusing phrasing that implies only a single vault can be used post
2.4

orthanc committed:
Document -client scripts for ansible vault.

orthanc committed:
Fix link to ask-vault-pass option

orthanc committed:
Add reference to multiple vault passwords into playbook vault docs

orthanc committed:
Fix phrasing of Since ansible 2.4

orthanc commented:
ready_for_review

bcoca commented on docs/docsite/rst/user_guide/playbooks_vault.rst:
```diff
@@ -7,26 +7,33 @@ Using Vault in playbooks
+.. note::
+    The :option:`--vault-id <ansible-playbook --vault-id>` flag is only avalibe in Ansible 2.4 or later.
+    Prior to Ansible 2.4, flags :option:`--ask-vault-pass <ansible-vault --ask-vault-pass>` or
+    :option:`--vault-password-file <ansible-vault --vault-password-file>` are used.
```
these are still usable if you dont want to use the 'id' system, i would leave them in the main text and maybe clarify this in the notes

bcoca commented on docs/docsite/rst/user_guide/playbooks_vault.rst:
```diff
@@ -37,9 +44,46 @@ If you are using a script instead of a flat file, ensure that it is marked as ex
+The :option:`--vault-id <ansible-pull --vault-id>` option can also be used with the :ref:`ansible-pull` command if you wish, though this would require distributing the keys to your nodes, so understand the implications -- vault is more intended for push mode.
+
+*Prior to Ansible 2.4*
+
```
the same here, this is in 'addition' as the old 'single vault options' still work

bcoca reviewed (dismissed):
Just don't imply that the 'single vault' options don't/won't work after 2.4, it should be clear for those using a single vault that they continue to work.

orthanc committed:
Tidy up of use of vault-id vs ask-vault-pass options in vault docs.

Use ask-vault-pass / vault-password-file whenever dealing with a single
unlabeled vault.

Introduce the concept of labeling a vault for clarity even if only a
single password is in use for a given run.

Rework multi-password section to align to these concepts.

orthanc committed:
Mention the new vault-id option

orthanc committed:
Mention client scripts in playbook vault docs

orthanc committed:
Associate the "for continuous integration" note with the
ANSIBLE_PASSWORD_FILE option.

orthanc commented:
ready_for_review

Based on discussion in the IRC chat it seems --ask-vault-pass and --vault-password-file are preferrable when there's only a single password in use.

So latest commits rework both the vault and playbookkk documentation to be consistent that:

* ask-vault-pass and vault-password-file are preferrable when there is only a single password in use
* labeled vault-id is a valid option for where there are multiple passwords in  use in the inventory but only one used in a given run
* labeled vault-id must be used when multiple passwords are needed for a single run
* unlabeled vault-id is valid, but rarely used (preferring ask-vault-pass or vault-password-file)

Then add the documentation of the client scripts in the context of that.

To make this read better I've moved the *Vault Ids and Multiple Vault Password* section of the vault docs to the top with the rest of the concepts so vault id's are introduced early before the examples of using them.

dglinder commented:
Thanks for the documentation update, @orthanc, around using multiple `--vault-id` entries on the `ansible-playbook` command, but I don't see anything covering how to create a variable with multiple versions using different `--vault-id`? Can you add some commentary around creating this using the `ansible-vault` encrypt options?

acozine commented on docs/docsite/rst/user_guide/playbooks_vault.rst:
```diff
@@ -7,20 +7,24 @@ Using Vault in playbooks
 For best practices advice, refer to :ref:`best_practices_for_variables_and_vaults`.
 
+.. note::
+    The :option:`--vault-id <ansible-playbook --vault-id>` flag is only avalibe in Ansible 2.4 or later.
```
typo: available not avalibe.

acozine replied:
Actually, can we remove this note entirely? We don't support anything older than 2.4 . . .

acozine commented on docs/docsite/rst/user_guide/playbooks_vault.rst:
```diff
@@ -30,16 +34,80 @@ Alternatively, passwords can be specified with a file or a script (the script ve
+    By defaut the vault label (dev, prod etc.) is just a hint. Ansible will try to decrypt each
+    vault with every provided password.
+
+    Setting the config option :ref:`DEFAULT_VAULT_ID_MATCH` will change this behaiour so that each password
```
typo: behavior not behaiour

acozine reviewed (commented):
I have not tested all the approaches described in this PR, but assuming that they are all correct, I'd say this PR is an improvement. Take care of the typos and it gets a thumbs-up from me.

orthanc committed:
Remove explicit Notes about lack of vault-id support before 2.4

Since prior to 2.4 is no longer supported these are redundant.

Also fix a typo.

orthanc commented:
ready_for_review

I've fixed the typos and removed the two explicit note boxes about support being added in 2.4.

There are a number of other references to 2.3, 2.4 or even 2.1 remaining in these two files, but replacing all of those requires a bit more re-writing so I'd prefer to do that in a subsequent PR.

Similarly I think @dglinder 's suggestion of clarifying creation of multiple vaults would also be better in a subequent PR as it will likely open a best practice / preference discussion. There's pretty resonable overap with the sections that need version number's removed so I'll likely tackle both of these as one PR.

bendem commented:
This makes ansible vaults really hard to use. The behavior is very counter-intuitive and without documentation, users are left digging the source code to understand how ansible treats vault-id scripts.
It took me a day to write a client script while having the documentation would have made it a 10 minutes task.

cabral404 commented:
I've never used multiple vault passwords so can't review that, but the rest looks good.

alongchamps commented:
Looks good to me and increases the number of examples.

bendem commented:
This is a no brainer, even if the documentation isn't perfect, it is still better than having nothing at all. Why isn't this being merged?

acozine commented on docs/docsite/rst/user_guide/playbooks_vault.rst:
```diff
@@ -30,16 +31,80 @@ Alternatively, passwords can be specified with a file or a script (the script ve
+If multiple vault passwords are required for a single run, :option:`--vault-id <ansible-playbook --vault-id>` must
+be used as it can be specified multiple times to provide the multiple passwords.  For example:
+
+To the 'dev' password from a file and prompt for the 'prod' password:
```
```suggestion
To read the 'dev' password from a file and prompt for the 'prod' password:
```

acozine commented on docs/docsite/rst/user_guide/playbooks_vault.rst:
```diff
@@ -30,16 +31,80 @@ Alternatively, passwords can be specified with a file or a script (the script ve
+the passwords, but it's generally cleaner to avoid mixing these with :option:`--vault-id <ansible-playbook --vault-id>`.
+
+.. note::
+    By defaut the vault label (dev, prod etc.) is just a hint. Ansible will try to decrypt each
```
```suggestion
    By default the vault label (dev, prod etc.) is just a hint. Ansible will try to decrypt each
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -27,6 +27,29 @@ given as the ``src`` argument to the :ref:`copy <copy_module>`, :ref:`template <
 
+.. _vault_ids:
+
+Vault Ids and Multiple Vault Passwords
```
```suggestion
Vault IDs and Multiple Vault Passwords
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -27,6 +27,29 @@ given as the ``src`` argument to the :ref:`copy <copy_module>`, :ref:`template <
+
+*Available since Ansible 2.4*
+
+A vault id is an identifier for one or more vault secrets. Since Ansible 2.4,
```
```suggestion
A vault ID is an identifier for one or more vault secrets. Since Ansible 2.4,
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -27,6 +27,29 @@ given as the ``src`` argument to the :ref:`copy <copy_module>`, :ref:`template <
+*Available since Ansible 2.4*
+
+A vault id is an identifier for one or more vault secrets. Since Ansible 2.4,
+Ansible supports multiple vault passwords. Vault ids is a way to provide
```
```suggestion
Ansible supports multiple vault passwords. Vault IDs provide
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -27,6 +27,29 @@ given as the ``src`` argument to the :ref:`copy <copy_module>`, :ref:`template <
+
+A vault id is an identifier for one or more vault secrets. Since Ansible 2.4,
+Ansible supports multiple vault passwords. Vault ids is a way to provide
+a label for a particular vault password.
```
```suggestion
labels for individual vault passwords.
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -27,6 +27,29 @@ given as the ``src`` argument to the :ref:`copy <copy_module>`, :ref:`template <
+Ansible supports multiple vault passwords. Vault ids is a way to provide
+a label for a particular vault password.
+
+Vault encrypted content can specify which vault id it was encrypted with.
```
```suggestion
Vault-encrypted content can specify which vault ID it was encrypted with.
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -27,6 +27,29 @@ given as the ``src`` argument to the :ref:`copy <copy_module>`, :ref:`template <
+passwords can be used at the same time.
+
+For example, a playbook can now include a vars file encrypted with a 'dev' vault
+id and a 'prod' vault id.
```
```suggestion
ID and a 'prod' vault ID.
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
+Labelling Vaults
+^^^^^^^^^^^^^^^^
+
+Since Ansible 2.4 the :option:`--vault-id <ansible-playbook --vault-id>` can be used to indicate which vault id
```
```suggestion
Since Ansible 2.4 the :option:`--vault-id <ansible-playbook --vault-id>` can be used to indicate which vault ID
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
+('dev', 'prod', 'cloud', etc) a password is for as well as how to source the password (prompt, a file path, etc).
+
+By default the vault-id label is only a hint, any values encrypted with the password will be decrypted.
+The config option :ref:`DEFAULT_VAULT_ID_MATCH` can be set to require the vault id to match the vault id
```
```suggestion
The config option :ref:`DEFAULT_VAULT_ID_MATCH` can be set to require the vault id to match the vault ID
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
     ansible-playbook --vault-id dev@dev-password site.yml
 
-To prompt for the 'dev' vault id:
+To prompt for the password for the 'dev' vault id:
```
```suggestion
To prompt for the password for the 'dev' vault ID:
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
     ansible-playbook --vault-id dev@prompt site.yml
 
-*Prior to Ansible 2.4*
+To get the 'dev' vault id password from an executable script :file:`my-vault-password.py`:
```
```suggestion
To get the 'dev' vault ID password from an executable script :file:`my-vault-password.py`:
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
+    ansible-playbook --vault-id dev@my-vault-password.py
 
-To be prompted for a vault password, use the :option:`--ask-vault-pass <ansible-playbook --vault-id>` cli option:
+Since Ansible 2.5 a single executable script can be used to get different password for depending on the vault label.  Scripts with
```
This doesn't scan quite right - "to get different password for depending on X" - but I'm not sure exactly what you meant here. Maybe: "Since Ansible 2.5 a single executable script can use vault IDs to get multiple passwords."?

orthanc replied:
You're right, this is confusing because it's really about the implementation of the script but it's mixed in with the "how to tell ansible which script to use" section.

I've re-written this part and broken it out into it's own section after the Multiple Vault Passwords section.

Hopefully it's clearer now.

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
+    contrib/vault/vault-keyring-client.py --vault-id dev
+
+
+The config option :ref:`DEFAULT_VAULT_IDENTITY_LIST` can be used to specify a default vault id and password source
```
```suggestion
The config option :ref:`DEFAULT_VAULT_IDENTITY_LIST` can be used to specify a default vault ID and password source
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
 
-In the above case, the 'dev' password will be tried first, then the 'prod' password for cases
-where Ansible doesn't know which vault id is used to encrypt something.
+By default the vault id label (dev, prod etc.) are only hints, Ansible will attempt to decrypt vault content
```
```suggestion
By default the vault ID labels (dev, prod etc.) are only hints, Ansible will attempt to decrypt vault content
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
-the provided vault ids. But the default is to try the matching id first, then try the other
-vault ids in order.
+In the above case, the 'dev' password will be tried first, then the 'prod' password for cases
+where Ansible doesn't know which vault id is used to encrypt something.
```
```suggestion
where Ansible doesn't know which vault ID is used to encrypt something.
```

acozine commented on docs/docsite/rst/user_guide/vault.rst:
```diff
@@ -196,120 +219,146 @@ Result::
 
-There is also a config option (:ref:`DEFAULT_VAULT_IDENTITY_LIST`) to specify a default list of vault ids to
-use. For example, instead of requiring the cli option on every use, the (:ref:`DEFAULT_VAULT_IDENTITY_LIST`) config option can be used:
+To add a vault id label to the encrypted data use the :option:`--vault-id <ansible-vault --vault-id>` option
```
```suggestion
To add a vault ID label to the encrypted data use the :option:`--vault-id <ansible-vault --vault-id>` option
```

acozine commented:
@bendem @alongchamps @cabral404 thanks for the reviews - it really helps to know which PRs are the most important to the community. I gave this a closer read, made some suggestions, and will ride herd on it today. I'm hoping @orthanc is in a time zone where the weekend hasn't begun yet!

acozine committed:
Apply typo fixes from code review

Co-Authored-By: orthanc <orthanc@users.noreply.github.com>

orthanc committed:
Split the explanation of Vault client scripts from vault usage since
it's about the implementation of the script rather than the usage.

orthanc committed:
Fix Title Case of Multiple Vault Passwords

orthanc committed:
Fix reference from Playbook Vault docs to Client Scripts and adopt
consistent heading for this topic between the two docs.

orthanc committed:
Consistent title case of Multiple Vault Passwords in playbook_vault docs to match vault docs

orthanc commented:
ready_for_review

I've accepted all @acozine 's excellent copy editing and split out the explination of Vault Client Scripts since it's really about the implementation of the script not how it's invoked.

bendem commented:
🎉
